### PR TITLE
8284949: riscv: Add Zero support for the 32-bit RISC-V architecture

### PR DESCRIPTION
**Pull request by feilongjiang**
This patch adds Zero support for the 32-bit RISC-V architecture.

Additional tests:

- [x] Linux zero RISCV32 cross-compilation
- [x] Resulting binaries run on QEMU User mode without problems

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284949](https://bugs.openjdk.java.net/browse/JDK-8284949): riscv: Add Zero support for the 32-bit RISC-V architecture

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to a26633b2c218a80dbcc02da6bc38955b76073e8e
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to dae310f94f73385b2d6747cf9738d641550e2807
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to dae310f94f73385b2d6747cf9738d641550e2807
 * [Yadong Wang](https://openjdk.java.net/census#yadongwang) (@yadongw - Author)

### Contributors
 * Junfeng Xie `<xiejunfeng3@huawei.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8284/head:pull/8284` \
`$ git checkout pull/8284`

Update a local copy of the PR: \
`$ git checkout pull/8284` \
`$ git pull https://git.openjdk.java.net/jdk pull/8284/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8284`

View PR using the GUI difftool: \
`$ git pr show -t 8284`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8284.diff">https://git.openjdk.java.net/jdk/pull/8284.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
Build change looks good.

File: src/hotspot/os/linux/waitBarrier_linux.cpp [hunk 1]
```diff
@@ -38,6 +38,13 @@
 
 #define guarantee_with_errno(cond, msg) check_with_errno(guarantee, cond, msg)
 
+// 32-bit RISC-V has no SYS_futex syscall.
+#ifndef SYS_futex
+  #if defined(RISCV32) && defined(SYS_futex_time64)
+    #define SYS_futex SYS_futex_time64
```
Comment (yadongw) [inline src/hotspot/os/linux/waitBarrier_linux.cpp:44]: Should it be better to check SYS_futex and SYS_futex_time64 when RISCV32 defined?

Comment (feilongjiang) [reply src/hotspot/os/linux/waitBarrier_linux.cpp:44]: Fixed.

**Review by tstuefe — APPROVED** [review]
LGTM

Cheers, Thomas

P.S. I assume you did not run the whole gamut of jtreg tests, right? Seems to me there are a number of tests which would need to get adapted for riscv32.

**Review by magicus — APPROVED** [review]

**Comment by feilongjiang** [comment]
> LGTM
> 
> Cheers, Thomas
> 
> P.S. I assume you did not run the whole gamut of jtreg tests, right? Seems to me there are a number of tests which would need to get adapted for riscv32.

Thanks for the review, Thomas. Currently, we only test jtreg tier1 tests under test/hotspot and test/jdk. I assume it would be enough for adding riscv32 zero support? There are some tests adaptions to be done for riscv32 like [1] [2]. I guess it might be better to fix them in another PR.

[1] https://github.com/openjdk/jdk/blob/72726c41829b33fd2baf5b3604cab49d39489dd2/test/lib/jdk/test/lib/Platform.java#L200-L202 
[2] https://github.com/openjdk/jdk/blob/72726c41829b33fd2baf5b3604cab49d39489dd2/test/lib-test/jdk/test/lib/TestMutuallyExclusivePlatformPredicates.java#L46-L49

**Review by yadongw — APPROVED** [review]

**Pull request closed without merging**